src/modules/abstraction-reform/lib/form-generator.js

```diff
@@ -0,0 +1,131 @@
+const { URL } = require('url');
+const refParser = require('json-schema-ref-parser');
+const { isObject, each } = require('lodash');
+const sentenceCase = require('sentence-case');
+const apiHelpers = require('./api-helpers');
+const { formFactory, fields } = require('../../../lib/forms');
+
+const types = {
+  ngr: require('../schema/types/ngr.json')
+};
+
+/**
+ * Given a picklist record and an array of picklist items loaded from the
+ * water service, generates a JSON schema for this picklist data
+ * @param {Object} picklist - picklist object from water service
+ * @param {Array} items - picklist items from water service
+ * @return {Object} JSON schema
+ */
+const picklistSchemaFactory = (picklist, items) => {
+  if (picklist.id_required) {
+    return {
+      type: 'object',
+      enum: items.map(item => ({ id: item.id, value: item.value }))
+    };
+  } else {
+    return {
+      type: 'string',
+      enum: items.map(item => item.value)
+    };
+  }
+};
+
+/**
+ * A resolver for json-schema-ref-parser
+ * This allows refs to be parsed and the relevant schema data to be
+ * retrieved
+ */
+const waterResolver = {
+  order: 1,
+
+  canRead: /^water:\/\//i,
+
+  read: async function (file) {
+    // Parse URL
+    const { host, pathname } = new URL(file.url);
+
+    const ref = pathname.replace('/', '').replace('.json', '');
+
+    // Load picklist schema from API
+    if (host === 'picklists') {
+      const picklist = await apiHelpers.getPicklist(ref);
+      const items = await apiHelpers.getPicklistItems(ref);
+
+      return picklistSchemaFactory(picklist, items);
+    }
+
+    // Load types from local files
+    if (host === 'types') {
+      return types[ref];
+    }
+  }
+
+};
+
+/**
+ * Converts references in the schema to their literals
+ * @param {Object} schema - JSON schema with refs
+ * @return {Promise} resolves with JSON schema with references converted to literals
+ */
+const dereference = async (schema) => {
+  const populated = await refParser.dereference(schema, { resolve: { waterResolver } });
+  return populated;
+};
+
+/**
+ * Guess label given a field name
+ * Converts underscore to space and converts to sentence case
+ * @param {String} str - field name, snake case
+ * @return {String} label
+ */
+const guessLabel = (str) => {
+  return sentenceCase(str.replace(/_/, ' '));
```

Comment:
Should this be `return sentenceCase(str.replace(/_/g, ' '));` to replace all the underscores?